src/listen/ldap_sync/proto_ldap_sync_ldap.c

```diff
@@ -0,0 +1,1325 @@
+ *
+ * @param[in,out] config	to check for attribute.
+ * @param[in] attr		to look for.
+ * @return
+ *	- 1 if attr is added
+ *	- 0 if attr was already present
+ */
+int ldap_sync_conf_attr_add(sync_config_t *config, char const * attr)
+{
+	char	**tmp;
+	size_t	len;
+
+	if (fr_ldap_attrs_check(config->attrs, attr)) return 0;
+
+	len = talloc_array_length(config->attrs);
+
+	config->attrs[len - 1] = talloc_strdup(config, attr);
+	tmp = (char **)talloc_array_null_terminate(UNCONST(void **, config->attrs));
+	memcpy(&config->attrs, &tmp, sizeof(config->attrs));
+
+	return 1;
+}
+
+static void _proto_ldap_socket_init(fr_connection_t *conn, UNUSED fr_connection_state_t prev,
+				    UNUSED fr_connection_state_t state, void *uctx);
+
+static void _proto_ldap_socket_open_connected(fr_connection_t *conn, UNUSED fr_connection_state_t prev,
+					      UNUSED fr_connection_state_t state, void *uctx);
+
+/** Attempt to (re)initialise a connection
+ *
+ * Performs complete re-initialization of a connection.  Called during socket_open
+ * to create the initial connection and again any time we need to reopen the connection.
+ *
+ * @param[in] el	the event list managing listen event.
+ * @param[in] now	current time.
+ * @param[in] user_ctx	Listener.
+ */
+static void proto_ldap_connection_init(UNUSED fr_event_list_t *el, UNUSED fr_time_t now, void *user_ctx)
+{
+	fr_listen_t			*listen = talloc_get_type_abort(user_ctx, fr_listen_t);
+	proto_ldap_sync_ldap_thread_t	*thread = talloc_get_type_abort(listen->thread_instance, proto_ldap_sync_ldap_thread_t);
+	proto_ldap_sync_ldap_t const	*inst = talloc_get_type_abort_const(thread->inst, proto_ldap_sync_ldap_t);
+
+	if (thread->conn) talloc_free(thread->conn);
+
+	/*
+	 *	Allocate an outbound LDAP connection
+	 */
+	thread->conn = fr_ldap_connection_state_alloc(thread, thread->el, &inst->handle_config, "ldap_sync");
+
+	if (!thread->conn) {
+		PERROR("Failed (re)initialising connection, will retry in %pV seconds",
+		       fr_box_time_delta(inst->handle_config.reconnection_delay));
+
+		if (fr_event_timer_in(thread, thread->el, &thread->conn_retry_ev,
+				      inst->handle_config.reconnection_delay,
+				      proto_ldap_connection_init, listen) < 0) {
+			FATAL("Failed inserting event: %s", fr_strerror());
+		}
+	}
+
+	/*
+	 *	Add watch functions on the LDAP connection
+	 */
+	fr_connection_add_watch_post(thread->conn, FR_CONNECTION_STATE_INIT,
+				     _proto_ldap_socket_init, true, thread);
+
+	fr_connection_add_watch_post(thread->conn, FR_CONNECTION_STATE_CONNECTED,
+				     _proto_ldap_socket_open_connected, true, thread);
+
+	/*
+	 *	Signal the connection to start
+	 */
+	fr_connection_signal_init(thread->conn);
+
+	return;
+}
+
+/** Dummy mod_open for child listener
+ *
+ * Openning of the real conneciton is done before this is called, but a mod_open
+ * is required in order to register the mod_read and mod_write.
+ */
+static int proto_ldap_child_mod_open(UNUSED fr_listen_t *li)
```

Comment:
Can you fix the I/O code to check for the presence of this callback and remove the dummy code